index.bs

```diff
@@ -165,6 +169,92 @@ To <dfn abstract-op lt="set-dest">set the `Sec-Fetch-Dest` header</dfn> for a [=
+servers SHOULD ignore this header if it contains an invalid value.
+
+<div algorithm="set `Sec-Fetch-Frame-Ancestors`">
+To <dfn abstract-op lt="set-frame-ancestors">set the `Sec-Fetch-Frame-Ancestors` header</dfn> for a [=request=] |r|:
+
+<ol class="algorithm">
+    1.  Assert: |r|'s [=request/url=] is a [=potentially trustworthy URL=].
+
+    1.  If |r|'s [=request/destination=] is "`document`", return.
+
+    1.  Let |header| be a [=Structured Header=] whose value is a [=structured header/token=].
+
+    1.  Set |header|'s value to "`same-origin`"..
+
+    1.  For each |ancestor| in |r|'s [=request/client=]'s [=environment settings object/global object=]'s
+        [=associated Document|associated `Document`=]'s [=Document/inclusive ancestor navigables=]:            
+
+        1.  Let |ancestor origin| be |ancestor|'s [=navigable/active document=]'s [=Document/origin=].
+
+        1.  If |ancestor origin| is not [=same site=] with |r|'s [=request/origin=], set |header|'s value
+            to "`cross-site`", and [=iteration/break=].
+
+        1.  If |ancestor origin| is not [=same origin=] with |r|'s [=request/origin=], set |header|'s value
+            to "`same-site`".
+
+    1.  [=header list/Set a structured field value=]
+        &#96;<a http-header>`Sec-Fetch-Frame-Ancestors`</a>&#96;/|header| in |r|'s [=request/header list=].
+</ol>
+</div>
+
+
+The `Sec-Fetch-Frame-Top` HTTP Request Header {#Sec-Fetch-Frame-Top}
+---------------------------------------------------------------------------------------
+
+The <dfn http-header export>`Sec-Fetch-Frame-Top`</dfn> HTTP request header exposes the
+relationship between a [=request=]'s target origin, and the origin of its top-frame document.
+It is a [=Structured Header=] whose value is a [=structured header/token=]. [[!I-D.ietf-httpbis-header-structure]] Its ABNF is:
+
+```
+Sec-Fetch-Frame-Top = sh-token
+```
+
+Valid `Sec-Fetch-Frame-Top` values include "`cross-site`", "`same-origin`", and "`same-site`".
+In order to support forward-compatibility with as-yet-unknown request types,
+servers SHOULD ignore this header if it contains an invalid value.
+
+Note: This header is delivered for all requests other than top-level navigations. Those are
+identifiable by examining the &#96;<a http-header>`Sec-Fetch-Dest`</a>&#96;header's value, so we
+can safely omit this header as it would be fully redundant in that case.
+
+<div algorithm="set `Sec-Fetch-Frame-Top`">
+To <dfn abstract-op lt="set-frame-top">set the `Sec-Fetch-Frame-Top` header</dfn> for a [=request=] |r|:
+
+<ol class="algorithm">
+    1.  Assert: |r|'s [=request/url=] is a [=potentially trustworthy URL=].
+
+    1.  If |r|'s [=request/destination=] is "`document`", return.
+
+    1.  Let |header| be a [=Structured Header=] whose value is a [=structured header/token=].
+
+    1.  Set |header|'s value to "`cross-site`".
+
+    1.  Let |top-frame origin| be the [=Document/top-level traversable=]'s [=Document/origin=].
```

Comment:
We need to specify *which* top-level traversable, since the UA may have multiple that exist simultaneously. I assume there's a way to get it from the request `|r|`, probably through its `client` -> `active document`?